examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3785 @@
+    let params: TransferCCDParams = ctx.parameter_cursor().get()?;
+
+    let amount = Amount::from_micro_ccd(params.amount.into());
+
+    // Send CCDs to receiver.
+    match params.address {
+        Receiver::Account(address) => host.invoke_transfer(&address, amount)?,
+        Receiver::Contract(address, function) => {
+            host.invoke_contract(&address, &params.data, function.as_entrypoint_name(), amount)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// This function is meant for recovering CCD tokens by the admin
+/// in case they accidentally end up on the implementation.
+/// Every smart contract instance with a payable function should have
+/// some mechanism to recover funds as a good smart contract coding practice.
+/// Nonetheless, it is not expected that any CCD tokens can end up
+/// on the implementation since the `wrap` function forwards them to the proxy.
+#[receive(
+    contract = "CIS2-wCCD",
+    parameter = "TransferCCDParams",
+    error = "ContractError",
+    name = "transferCCD",
+    mutable
+)]
+fn contract_implementation_transfer_ccd<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<()> {
+    // Check that only the admin is authorized.
+    ensure_eq!(ctx.sender(), host.state().admin, ContractError::Unauthorized);
+
+    let params: TransferCCDParams = ctx.parameter_cursor().get()?;
+
+    let amount = Amount::from_micro_ccd(params.amount.into());
+
+    match params.address {
+        Receiver::Account(address) => host.invoke_transfer(&address, amount)?,
+        Receiver::Contract(address, function) => {
+            host.invoke_contract(&address, &params.data, function.as_entrypoint_name(), amount)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Function to view contract balance of state contract.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "viewBalance",
+    return_value = "Amount",
+    error = "ContractError"
+)]
+fn contract_state_view_balance<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<Amount> {
+    Ok(host.self_balance())
+}
+
+/// Function to view contract balance of implementation contract.
+#[receive(
+    contract = "CIS2-wCCD",
+    name = "viewBalance",
+    return_value = "Amount",
+    error = "ContractError"
+)]
+fn contract_view_balance<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<StateImplementation, StateApiType = S>,
+) -> ContractResult<Amount> {
+    Ok(host.self_balance())
+}
+
+/// Function to view contract balance of proxy contract.
+#[receive(
+    contract = "CIS2-wCCD-Proxy",
+    name = "viewBalance",
+    return_value = "Amount",
+    error = "ContractError"
+)]
+fn contract_proxy_view_balance<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<StateProxy, StateApiType = S>,
+) -> ContractResult<Amount> {
+    Ok(host.self_balance())
+}
```

Comment:
These functions are not used internally and if you want to know the balance of a smart contract off-chain, you can use ``concordium-client``. 

But I see an use case that smart contracts can query the balance of any other smart contract within the Rust code (e.g. `(smartContractAddress).balance` this is currently not natively supported but on the smart contract wish list for future development). 

Since it is not available at the moment, I added a view function in case a third-party smart contract needs this information.

Reply:
I don't see why other smart contracts would need this. Especially not for the internal contracts such as implementation and state. 
I think we should be prioritizing having a smaller module instead.

Reply:
I don't know any concrete example of why a third-party smart contract would want to query the balance of the wCCD smart contract.

Since we could do a protocol update one day and add this `(smartContractAddress).balance` feature, I can remove the above addresses, and we prioritize the module size now.